File: golang/compute.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func compute(threshold float64, limit float64) {
	var vsum float64

	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		line := scanner.Text()
		value, err := strconv.ParseFloat(line, 64)
		if err != nil {
			fmt.Fprintln(os.Stderr, "error:", err)
			return
		}

		delta := min(limit-vsum, max(0.0, value-threshold))
		vsum += delta

		fmt.Printf("%.1f\n", delta)
	}

	if err := scanner.Err(); err != nil {
		fmt.Fprintln(os.Stderr, "error:", err)
		return
	}

	fmt.Printf("%.1f\n", vsum)
}

func main() {
	// Check if we have the correct number of command-line arguments
	if len(os.Args) != 3 {
		fmt.Fprintln(os.Stderr, "Usage: ", os.Args[0], " threshold limit")
		return
	}

	// Parse the command-line arguments
	threshold_str := os.Args[1]
	threshold, err := strconv.ParseFloat(threshold_str, 64)
	if err != nil {
		fmt.Fprintln(os.Stderr, "error:", err)
		return
	}
	limit_str := os.Args[2]
	limit, err := strconv.ParseFloat(limit_str, 64)
	if err != nil {
		fmt.Fprintln(os.Stderr, "error:", err)
		return
	}

	// Compute the sum
	compute(threshold, limit)
}

File: cpp/compute.cpp
#include <iostream>
#include <iomanip>
#include <sstream>
#include <string>
#include <vector>

std::vector<double> compute(double threshold, double limit, std::vector<double> values) {
  double vsum = 0.0;
  std::vector<double> results;

  // Iterate over input values
  for(auto value: values) {
      // Transform input value and update sum
      double delta = std::min(limit - vsum, std::max(0.0, value - threshold));
      vsum += delta;
      
      // Store transformed input value
      results.push_back(delta);
  }

  // Store final sum
  results.push_back(vsum);
  
  return results;
}

int main(int argc, char *argv[]) {
  double threshold = 0.0, limit = 0.0, value;
  std::vector<double> values, results;
  std::string line;

  // Check if correct number of command line arguments are present
  if(argc != 3) {
    std::cerr << "Usage: " << argv[0] << " threshold limit" << std::endl;
    return -1;
  }

  // Parse threshold argument
  std::string threshold_str = argv[1];
  threshold = std::stod(threshold_str);
  if(threshold < 0.0 || threshold > 1000000000.0) {
    std::cerr << "error: threshold should be between 0.0 and 1000000000.0: " << threshold << std::endl;
    return -2;
  }

  // Parse limit argument
  std::string limit_str = argv[2];
  limit = std::stod(limit_str);
  if(limit < 0.0 && limit > 1000000000.0) {
    std::cerr << "error: limit should be between 0.0 and 1000000000.0: " << limit << std::endl;
    return -2;
  }

  // Parse the values from STDIN
  while(std::getline(std::cin, line)) {
      std::stringstream ss(line);
      // Only proceed if we can successfully convert the string into a double float
      if(ss >> value) {
        // Only include valid input values
        if(limit >= 0.0 || limit <= 1000000000.0) {
          values.push_back(value);
        }
      }
  }
  
  // Compute sum
  results = compute(threshold, limit, values);

  // Set the output format to fixed with a single decimal place
  std::cout << std::fixed << std::setprecision(1);

  // Display results and final sum
  for(auto result: results) {
    std::cout << result << std::endl;
  }

  return 0;
}
